arraycontext/container/traversal.py

```diff
@@ -520,6 +522,71 @@ def to_numpy(ary: Any, actx: ArrayContext) -> Any:
     """
     return rec_map_array_container(actx.to_numpy, ary)
 
+
+def flatten_to_numpy(ary: ArrayOrContainerT, actx: ArrayContext) -> np.ndarray:
+    """Convert all arrays in the :class:`~arraycontext.ArrayContainer`
+    to host :mod:`numpy` arrays, flatten them using :func:`~numpy.ravel`
+    and concatenate them into a single :class:`~numpy.ndarray`.
+
+    The order in which the individual leaf arrays appear in the final array is
+    dependent on the order given by :func:`~arraycontext.serialize_container`.
+    """
+    def _flatten_to_numpy(subary: ArrayOrContainerT) -> None:
+        try:
+            iterable = serialize_container(subary)
+        except TypeError:
+            result.append(actx.to_numpy(subary).ravel())
+        else:
+            for _, isubary in iterable:
+                _flatten_to_numpy(isubary)
+
+    result: List[np.ndarray] = []
+    _flatten_to_numpy(ary)
+
+    return np.concatenate(result)
+
+
+def unflatten_from_numpy(
+        template: ArrayOrContainerT, ary: np.ndarray,
+        actx: ArrayContext) -> ArrayOrContainerT:
+    """Unflatten an :class:`~numpy.ndarray` produced by :func:`flatten_to_numpy`
+    back into an :class:`~arraycontext.ArrayContainer`.
+
+    The order and sizes of each slice into *ary* are determined by the
+    array container *template*.
+    """
+    # NOTE: https://github.com/python/mypy/issues/7057
+    offset = 0
+
+    def _unflatten_from_numpy(subary: ArrayOrContainerT) -> ArrayOrContainerT:
+        nonlocal offset
+
+        try:
+            iterable = serialize_container(subary)
```

Comment:
FWIW: I tried these functions out in `grudge` for PR https://github.com/inducer/grudge/pull/154 and they work like a charm! I tested this on several ranks and I don't observe any issues.

Reply:
Hm, yeah, I can see this failing in a distributed case if the groups on both sides decide to number things differently.

I'm not quite sure how to fix something like that. It seems like it would need a bit more metadata about the tree structure, so it might be easier to teach array containers to pickle themselves and just send that over the wire?

Reply:
@alexfikl One way that I attempted to rectify this in https://github.com/inducer/grudge/pull/154 was to try to pass an ordered list of keys that were traversed in `serialize_container` _before_ flattening, then using that order to compute the right offset parameter when performing unflattering.

I'll try to sketch something out for this PR

Reply:
Would that help though? That handles the order on one side, but the other side in the send/recv might have a different order, right?

Reply:
Or array containers need to be "ordered" in some logical way? Not sure what that means exactly for our case.

Reply:
> Would that help though? That handles the order on one side, but the other side in the send/recv might have a different order, right?

Yeah I updated my comment. I realize now that likely won't help :cry:

Reply:
At least for the array containers we have, they're all consistent in the sense that they always give the same result when serialized. E.g.
* the `DOFArray` will always serialize the groups in the order they were given to it.
* the dataclass-based ones will always serialize the attributes in the order they're written in the class definition.

So the container should "know" its ordering, even if we don't impose one. Am I misunderstanding what order you mean?

Reply:
In which case, perhaps we don't necessarily have a problem in the first place? Meaning what is currently implemented is "safe". That would explain why I don't appear to be having issues communicating data across ranks in https://github.com/inducer/grudge/pull/154 when using this PR branch

Reply:
Hm, yeah, if it works in practice, that's big :+1:!

It still feels like there's a potential for something fishy to happen, but I don't know enough about how meshmode does the partitioning to come up with an example where this would fail somehow.

Reply:
See also #92.

Reply:
Added a sentence to the docs to mention that this is required in 17f74ad.

Let me know that's sufficient / phrased unambiguously enough.

Reply:
My worry in the distributed case was more for a scenario like this:
* take 2 ranks with 2 groups that didn't number the groups in the same way
* then they need to exchange the data in a cross, i.e. `(rank0, group0)` goes to `(rank1, group1)` and `(rank0, group1)` goes to `(rank1, group0)`. 

This would completely confuse the flattening as it is now. Is that something that can happen?

Reply:
I agree that that would completely confuse the unflattening, but I don't think there's a big risk. In the case of cross-rank communication, the boundary mesh from the sending rank is shipped to the receiver, and that then uses a connection to get the data onto its own discretization boundary discretization. The use case here is mainly one discretization talking to itself (even if it is on a different rank).

Reply:
Ah, I see, thanks for the explanation! 
Then I would say this is resolved with the little blurb in the `serialize_contianer` docs.